core/trace/dt_span_exporter.go

```diff
@@ -87,25 +86,25 @@ func (e *dtSpanExporterImpl) export(ctx context.Context, t exportType, spans dtS
 	e.logger.Debugf("Serialize %d spans to export", len(spans))
 
 	start := time.Now()
-	// TODO: In order to support large amounts of spans, implement a splitting algorithm
-	// so that we can send spans in batches whose sizes do not exceed cMaxSizeSend.
-	serializedSpans, err := serializeSpans(spans, e.config.Tenant, e.config.AgentId, e.config.QualifiedTenantId())
+	serializedSpanExports, err := e.serializer.serializeSpans(spans, e.config.Tenant, e.config.AgentId, e.config.QualifiedTenantId())
 	if err != nil {
 		return err
 	}
 
 	e.logger.Debugf("Serialization process took %s", time.Since(start))
 
-	serializedSpansLen := len(serializedSpans)
-	if serializedSpansLen > cMaxSizeSend {
-		errMsg := fmt.Sprintf("skip exporting, serialized spans reached %d bytes. Maximum allowed size is %d bytes",
-			serializedSpansLen, cMaxSizeSend)
-		return errors.New(errMsg)
-	} else if serializedSpansLen > cMaxSizeWarning {
-		e.logger.Warnf("Size of serialized spans reached %d bytes", serializedSpansLen)
+	for _, spanExport := range serializedSpanExports {
```

Comment:
i believe there is now spans x 2 memory usage which might be multiples of 64MB in the worst case

would it make sense to pass a callback to `serializeSpans` that which would get called for each serailized span and do the exporting there?

Reply:
That sounds like it would be possible, though I don't think it is done this way in any other language if I'm not mistaken.

Reply:
In Java we do that, see https://bitbucket.lab.dynatrace.org/projects/ODIN/repos/odin-java/browse/dynatrace/sdk/src/main/java/com/dynatrace/opentelemetry/SpanSerializer.java#103-138

The SpanSerializer is constructed from a SpanExport.Builder and an iterator over spans. The addNextSpansToProtoAndMarkBuffered method is then repeatedly called. It clears the Builder's span array and appends the next batch if any, returning false if it reached the end (without appending further spans).

This has implications not only for memory usage but also retry handling, as only the spans that we actually started sending are marked as buffered (like the spec calls it).